test/Microsoft.SqlTools.ServiceLayer.IntegrationTests/LanguageServer/LanguageServiceTests.cs

```diff
@@ -90,11 +97,57 @@ public void AutoCompleteFindCompletions()
             var completions = autoCompleteService.GetCompletionItems(
                 result.TextDocumentPosition,
                 result.ScriptFile,
-                result.ConnectionInfo);
+                result.ConnectionInfo).Result;
 
             Assert.True(completions.Length > 0);
         }
 
+        public static string AssemblyDirectory
+        {
+            get
+            {
+                string codeBase = Assembly.GetExecutingAssembly().CodeBase;
+                UriBuilder uri = new UriBuilder(codeBase);
+                string path = Uri.UnescapeDataString(uri.Path);
+                return Path.GetDirectoryName(path);
+            }
+        }
+
+        /// <summary>
+        /// This test tests completion extension interface
+        /// </summary>
+        [Fact]
+        public async void AutoCompleteWithExtension()
+        {
+            var result = GetLiveAutoCompleteTestObjects();
+
+            result.TextDocumentPosition.Position.Character = 10;
+            result.ScriptFile = ScriptFileTests.GetTestScriptFile("select * f");
+            result.TextDocumentPosition.TextDocument.Uri = result.ScriptFile.FilePath;
+
+            var autoCompleteService = LanguageService.Instance;
+            var requestContext = new Mock<SqlTools.Hosting.Protocol.RequestContext<bool>>();
+            requestContext.Setup(x => x.SendResult(It.IsAny<bool>()))
+                .Returns(Task.FromResult(new object()));
+            var extensionParams = new CompletionExtensionParams()
+            {
+                Assembly = Path.Combine(AssemblyDirectory, "Microsoft.SqlTools.Test.CompletionExtension.dll"),
+                TypeName = "Microsoft.SqlTools.Test.CompletionExtension.CompletionExtProvider1",
+                Properties = new Dictionary<string, object> { { "modelPath", "testModel" } }
+            };
+            await autoCompleteService.HandleCompletionExtLoadRequest(extensionParams, requestContext.Object);
+            ScriptParseInfo scriptInfo = new ScriptParseInfo { IsConnected = true };
+            autoCompleteService.ParseAndBind(result.ScriptFile, result.ConnectionInfo);
+            scriptInfo.ConnectionKey = autoCompleteService.BindingQueue.AddConnectionContext(result.ConnectionInfo);
+
+            var completions = autoCompleteService.GetCompletionItems(
+                result.TextDocumentPosition,
+                result.ScriptFile,
+                result.ConnectionInfo).Result;
+
+            Assert.True(completions.Length > 0 && completions[0].Preselect.HasValue && completions[0].Preselect.Value);
```

Comment:
>True [](start = 19, length = 4)

Add a message here explaining the issue - right now if this fails it's just going to say "expected true got false" which is not very helpful in figuring out the issue. Make sure you include the contents of the completions list as well since that often times helps with figuring out an issue.

Reply:
Added more comments to the test.

Reply:
Please add messages to the actual asserts as well - that way when they fail the log will have useful information instead of just saying "expected true got false"

---
In reply to: [304579406](https://github.com/microsoft/sqltoolsservice/pull/833#discussion_r304579406) [](ancestors = 304579406)

Reply:
I see, added.